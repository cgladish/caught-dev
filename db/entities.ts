export type ServiceAuthEntity = {
  id: number;
  appName: AppName;
  encryptedToken: string;
};

export type PreservationRuleEntity = {
  id: number;
  appName: AppName;
  name: string;
  selectedJson: string;
  startDatetime: string | null;
  endDatetime: string | null;
  createdAt: string;
  updatedAt: string;
};

export type MessageEntity = {
  id: number;
  preservationRuleId: number;
  externalId: string;
  externalChannelId: string;
  authorId: string;
  authorName: string;
  authorAvatar?: string;
  content: string;
  sentAt: string;
  appSpecificDataJson?: string;
};
