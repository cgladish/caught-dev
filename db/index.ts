import knex from "knex";
const config = require("./knexfile");

const environment = process.env.JEST_WORKER_ID ? "test" : "production";

const db = knex(config[environment]);
const migrate = db.migrate.latest(config[environment].migrations);

export default async function getDb() {
  await migrate;
  return db;
}
