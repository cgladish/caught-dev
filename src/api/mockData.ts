import { v4 } from "uuid";
import getDb from "../db";
import {
  ChannelEntity,
  MessageEntity,
  PreservationRuleEntity,
  WordCountEntity,
} from "../db/entities";
import TableName from "../db/tableName";

export const makeMessage = async (
  preservationRuleId: number,
  overwrites?: Partial<MessageEntity>
): Promise<MessageEntity> => {
  const db = await getDb();
  const [id] = await db<MessageEntity>(TableName.Message).insert({
    preservationRuleId,
    externalId: v4(),
    externalChannelId: "1234567",
    authorId: "1234567",
    authorName: "Name",
    content: "Content",
    sentAt: "2022-09-21T02:00:00.000Z",
    appSpecificDataJson: null,
    ...overwrites,
  });
  const entity = await db<MessageEntity>(TableName.Message)
    .where({ id })
    .first();
  return entity!;
};

export const makeWordCount = async (
  preservationRuleId: number,
  overwrites?: Partial<WordCountEntity>
): Promise<WordCountEntity> => {
  const db = await getDb();
  const [id] = await db<WordCountEntity>(TableName.WordCount).insert({
    preservationRuleId,
    word: "word",
    count: 5,
    ...overwrites,
  });
  const entity = await db<WordCountEntity>(TableName.WordCount)
    .where({ id })
    .first();
  return entity!;
};

export const makePreservationRule = async (
  overwrites?: Partial<PreservationRuleEntity>
): Promise<PreservationRuleEntity> => {
  const db = await getDb();
  const [id] = await db<PreservationRuleEntity>(
    TableName.PreservationRule
  ).insert({
    appName: "discord",
    name: v4(),
    selectedJson: '{"key":"value"}',
    startDatetime: "2022-09-21T02:00:00.000Z",
    endDatetime: "2022-09-21T04:00:00.000Z",
    initialBackupComplete: false,
    ...overwrites,
  });
  const entity = await db<PreservationRuleEntity>(TableName.PreservationRule)
    .where({ id })
    .first();
  return entity!;
};

export const makeChannel = async (
  overwrites?: Partial<ChannelEntity>
): Promise<ChannelEntity> => {
  const db = await getDb();
  const [id] = await db<ChannelEntity>(TableName.Channel).insert({
    appName: "discord",
    externalId: v4(),
    name: "channel name",
    iconUrl: null,
    ...overwrites,
  });
  const entity = await db<ChannelEntity>(TableName.Channel)
    .where({ id })
    .first();
  return entity!;
};
