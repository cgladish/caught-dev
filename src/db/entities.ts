export type ServiceAuthEntity = {
  id: number;
  appName: AppName;
  encryptedToken: string;
};

export type PreservationRuleEntity = {
  id: number;
  appName: AppName;
  name: string;
  selectedJson: string;
  startDatetime?: string | null;
  endDatetime?: string | null;
  initialBackupComplete: boolean;
  createdAt: string;
  updatedAt: string;
};

export type ChannelEntity = {
  id: number;
  appName: AppName;
  externalId: string;
  name: string;
  iconUrl?: string | null;
};

export type MessageEntity = {
  id: number;
  preservationRuleId: number;
  externalId: string;
  externalChannelId: string;
  authorId: string;
  authorName: string;
  authorAvatar?: string | null;
  content: string;
  sentAt: string;
  appSpecificDataJson?: string | null;
};

export type WordCountEntity = {
  id: number;
  preservationRuleId: number;
  word: string;
  count: number;
};
