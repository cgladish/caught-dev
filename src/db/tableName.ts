enum TableName {
  ServiceAuth = "ServiceAuth",
  PreservationRule = "PreservationRule",
  Message = "Message",
  MessageSearch = "MessageSearch",
  Channel = "Channel",
  WordCount = "WordCount",
}

export default TableName;
