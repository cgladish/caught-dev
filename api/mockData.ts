import { v4 } from "uuid";
import getDb from "../db";
import { MessageEntity, PreservationRuleEntity } from "../db/entities";
import TableName from "../db/tableName";

export const makeMessage = async (
  preservationRuleId: number,
  overwrites?: Partial<MessageEntity>
): Promise<MessageEntity> => {
  const db = await getDb();
  const [id] = await db<MessageEntity>(TableName.Message).insert({
    preservationRuleId,
    externalId: v4(),
    externalChannelId: "1234567",
    authorId: "1234567",
    authorName: "Name",
    content: "Content",
    sentAt: "2022-09-21T02:00:00.000Z",
    ...overwrites,
  });
  const entity = await db<MessageEntity>(TableName.Message)
    .where({ id })
    .first();
  return entity!;
};

export const makePreservationRule = async (
  overwrites?: Partial<PreservationRuleEntity>
): Promise<PreservationRuleEntity> => {
  const db = await getDb();
  const [id] = await db<PreservationRuleEntity>(
    TableName.PreservationRule
  ).insert({
    appName: "discord",
    name: v4(),
    selectedJson: '{"key":"value"}',
    startDatetime: "2022-09-21T02:00:00.000Z",
    endDatetime: "2022-09-21T04:00:00.000Z",
    initialBackupComplete: false,
    ...overwrites,
  });
  const entity = await db<PreservationRuleEntity>(TableName.PreservationRule)
    .where({ id })
    .first();
  return entity!;
};
