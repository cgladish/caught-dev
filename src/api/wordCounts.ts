import getDb from "../db";
import { MessageEntity, WordCountEntity } from "../db/entities";
import TableName from "../db/tableName";
import chunk from "lodash/chunk";
import { removeStopwords } from "stopword";

type TempWordCountEntity = Omit<WordCountEntity, "preservationRuleId">;

export const fetchTopWordCounts = async (
  preservationRuleId: number,
  limit: number = 200
): Promise<WordCountEntity[]> => {
  const db = await getDb();
  return db<WordCountEntity>(TableName.WordCount)
    .where({ preservationRuleId })
    .orderBy("count", "desc")
    .orderBy("word", "asc")
    .limit(limit);
};

const INSERT_BATCH_SIZE = 100;
export const updateWordCounts = async (
  preservationRuleId: number,
  messages: Pick<MessageEntity, "content">[]
): Promise<void> => {
  const wordCounts: Record<string, number> = {};
  messages.forEach(({ content }) => {
    const words = removeStopwords(
      content
        .replace(/[.,\/#!\?$%\^&\*;:{}=\-_`~()]/g, "")
        .replace(/\s{2,}/g, " ")
        .trim()
        .split(" ")
    );
    words.forEach((word) => {
      const wordLower = word.toLocaleLowerCase();
      wordCounts[wordLower] = (wordCounts[wordLower] ?? 0) + 1;
    });
  });
  const wordCountEntities: Omit<TempWordCountEntity, "id">[] = Object.entries(
    wordCounts
  ).map(([word, count]) => ({
    word,
    count,
  }));
  const db = await getDb();
  await db.transaction(async (tx) => {
    const tempTableName = `TempWordCounts${preservationRuleId}`;
    await tx.schema.createTable(tempTableName, (table) => {
      table.increments("id");
      table.string("word").notNullable().index();
      table.integer("count").notNullable();
    });
    for (let wordCountEntityChunk of chunk(
      wordCountEntities,
      INSERT_BATCH_SIZE
    )) {
      await tx
        .table<TempWordCountEntity>(tempTableName)
        .insert(wordCountEntityChunk);
    }

    await tx.raw(`
      UPDATE ${TableName.WordCount}
      SET count = a.count + b.count
      FROM ${TableName.WordCount} a
      INNER JOIN ${tempTableName} b ON a.word = b.word
      WHERE (
        a.preservationRuleId = ${preservationRuleId} AND
        ${TableName.WordCount}.id = a.id
      )
    `);

    const newWords: { word: string }[] = await tx.raw(`
      SELECT word
      FROM ${tempTableName}
      WHERE NOT EXISTS (
        SELECT id FROM ${TableName.WordCount}
        WHERE (
          ${TableName.WordCount}.preservationRuleId = ${preservationRuleId} AND
          ${TableName.WordCount}.word = ${tempTableName}.word
        )
      )
    `);
    const newWordCountEntities = wordCountEntities
      .filter(({ word }) => newWords.some((newWord) => newWord.word === word))
      .map((entity) => ({ ...entity, preservationRuleId }));

    for (let wordCountEntityChunk of chunk(
      newWordCountEntities,
      INSERT_BATCH_SIZE
    )) {
      await tx
        .table<WordCountEntity>(TableName.WordCount)
        .insert(wordCountEntityChunk);
    }

    await tx.schema.dropTable(tempTableName);
  });
};
