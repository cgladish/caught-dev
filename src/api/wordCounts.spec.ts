import { WordCountEntity } from "../db/entities";
import TableName from "../db/tableName";
import { Knex } from "knex";
import getDb from "../db";
import { makeMessage, makePreservationRule, makeWordCount } from "./mockData";
import { fetchTopWordCounts, updateWordCounts } from "./wordCounts";

describe("preservationRules", () => {
  let db: Knex;

  beforeEach(async () => {
    db = await getDb();
  });

  describe("fetchTopWordCounts", () => {
    it("successfully fetches top word counts", async () => {
      const { id: preservationRuleId } = await makePreservationRule();
      for (let i = 0; i < 10; ++i) {
        const wordCounts = new Array(100).fill(null).map((_, j) => ({
          preservationRuleId,
          word: `word-${i * 100 + j}`,
          count: i * 100 + j,
        }));
        await db<WordCountEntity>(TableName.WordCount).insert(wordCounts);
      }

      expect(await fetchTopWordCounts(preservationRuleId)).toMatchSnapshot();
    });
  });

  describe("updateWordCounts", () => {
    it("successfully updates word counts for a preservation rule", async () => {
      const { id: preservationRuleId } = await makePreservationRule();
      const { id: otherPreservationRuleId } = await makePreservationRule();
      const messages = await Promise.all([
        makeMessage(preservationRuleId, {
          content: "Hello, this is a test message!",
        }),
        makeMessage(preservationRuleId, { content: "Also a test message :)" }),
        makeMessage(preservationRuleId, {
          content: "Maybe not a test message?",
        }),
        makeMessage(preservationRuleId, {
          content: "Definitely a test  message!!",
        }),
      ]);
      await Promise.all([
        makeWordCount(preservationRuleId, { word: "hello", count: 2 }),
        makeWordCount(preservationRuleId, { word: "message", count: 1 }),
        makeWordCount(preservationRuleId, { word: "test", count: 3 }),
        makeWordCount(preservationRuleId, { word: "other", count: 6 }),
        makeWordCount(otherPreservationRuleId, { word: "test", count: 1 }),
      ]);

      await updateWordCounts(preservationRuleId, messages);

      const wordCounts = await db<WordCountEntity>(TableName.WordCount)
        .select("word", "count")
        .orderBy("word", "asc")
        .orderBy("count", "asc");
      expect(wordCounts).toMatchSnapshot();
    });
  });
});
