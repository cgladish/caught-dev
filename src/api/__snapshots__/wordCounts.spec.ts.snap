// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`preservationRules getTopWordCounts successfully fetches top word counts 1`] = `
[
  {
    "count": 999,
    "id": 1000,
    "preservationRuleId": 1,
    "word": "word-999",
  },
  {
    "count": 998,
    "id": 999,
    "preservationRuleId": 1,
    "word": "word-998",
  },
  {
    "count": 997,
    "id": 998,
    "preservationRuleId": 1,
    "word": "word-997",
  },
  {
    "count": 996,
    "id": 997,
    "preservationRuleId": 1,
    "word": "word-996",
  },
  {
    "count": 995,
    "id": 996,
    "preservationRuleId": 1,
    "word": "word-995",
  },
  {
    "count": 994,
    "id": 995,
    "preservationRuleId": 1,
    "word": "word-994",
  },
  {
    "count": 993,
    "id": 994,
    "preservationRuleId": 1,
    "word": "word-993",
  },
  {
    "count": 992,
    "id": 993,
    "preservationRuleId": 1,
    "word": "word-992",
  },
  {
    "count": 991,
    "id": 992,
    "preservationRuleId": 1,
    "word": "word-991",
  },
  {
    "count": 990,
    "id": 991,
    "preservationRuleId": 1,
    "word": "word-990",
  },
  {
    "count": 989,
    "id": 990,
    "preservationRuleId": 1,
    "word": "word-989",
  },
  {
    "count": 988,
    "id": 989,
    "preservationRuleId": 1,
    "word": "word-988",
  },
  {
    "count": 987,
    "id": 988,
    "preservationRuleId": 1,
    "word": "word-987",
  },
  {
    "count": 986,
    "id": 987,
    "preservationRuleId": 1,
    "word": "word-986",
  },
  {
    "count": 985,
    "id": 986,
    "preservationRuleId": 1,
    "word": "word-985",
  },
  {
    "count": 984,
    "id": 985,
    "preservationRuleId": 1,
    "word": "word-984",
  },
  {
    "count": 983,
    "id": 984,
    "preservationRuleId": 1,
    "word": "word-983",
  },
  {
    "count": 982,
    "id": 983,
    "preservationRuleId": 1,
    "word": "word-982",
  },
  {
    "count": 981,
    "id": 982,
    "preservationRuleId": 1,
    "word": "word-981",
  },
  {
    "count": 980,
    "id": 981,
    "preservationRuleId": 1,
    "word": "word-980",
  },
  {
    "count": 979,
    "id": 980,
    "preservationRuleId": 1,
    "word": "word-979",
  },
  {
    "count": 978,
    "id": 979,
    "preservationRuleId": 1,
    "word": "word-978",
  },
  {
    "count": 977,
    "id": 978,
    "preservationRuleId": 1,
    "word": "word-977",
  },
  {
    "count": 976,
    "id": 977,
    "preservationRuleId": 1,
    "word": "word-976",
  },
  {
    "count": 975,
    "id": 976,
    "preservationRuleId": 1,
    "word": "word-975",
  },
  {
    "count": 974,
    "id": 975,
    "preservationRuleId": 1,
    "word": "word-974",
  },
  {
    "count": 973,
    "id": 974,
    "preservationRuleId": 1,
    "word": "word-973",
  },
  {
    "count": 972,
    "id": 973,
    "preservationRuleId": 1,
    "word": "word-972",
  },
  {
    "count": 971,
    "id": 972,
    "preservationRuleId": 1,
    "word": "word-971",
  },
  {
    "count": 970,
    "id": 971,
    "preservationRuleId": 1,
    "word": "word-970",
  },
  {
    "count": 969,
    "id": 970,
    "preservationRuleId": 1,
    "word": "word-969",
  },
  {
    "count": 968,
    "id": 969,
    "preservationRuleId": 1,
    "word": "word-968",
  },
  {
    "count": 967,
    "id": 968,
    "preservationRuleId": 1,
    "word": "word-967",
  },
  {
    "count": 966,
    "id": 967,
    "preservationRuleId": 1,
    "word": "word-966",
  },
  {
    "count": 965,
    "id": 966,
    "preservationRuleId": 1,
    "word": "word-965",
  },
  {
    "count": 964,
    "id": 965,
    "preservationRuleId": 1,
    "word": "word-964",
  },
  {
    "count": 963,
    "id": 964,
    "preservationRuleId": 1,
    "word": "word-963",
  },
  {
    "count": 962,
    "id": 963,
    "preservationRuleId": 1,
    "word": "word-962",
  },
  {
    "count": 961,
    "id": 962,
    "preservationRuleId": 1,
    "word": "word-961",
  },
  {
    "count": 960,
    "id": 961,
    "preservationRuleId": 1,
    "word": "word-960",
  },
  {
    "count": 959,
    "id": 960,
    "preservationRuleId": 1,
    "word": "word-959",
  },
  {
    "count": 958,
    "id": 959,
    "preservationRuleId": 1,
    "word": "word-958",
  },
  {
    "count": 957,
    "id": 958,
    "preservationRuleId": 1,
    "word": "word-957",
  },
  {
    "count": 956,
    "id": 957,
    "preservationRuleId": 1,
    "word": "word-956",
  },
  {
    "count": 955,
    "id": 956,
    "preservationRuleId": 1,
    "word": "word-955",
  },
  {
    "count": 954,
    "id": 955,
    "preservationRuleId": 1,
    "word": "word-954",
  },
  {
    "count": 953,
    "id": 954,
    "preservationRuleId": 1,
    "word": "word-953",
  },
  {
    "count": 952,
    "id": 953,
    "preservationRuleId": 1,
    "word": "word-952",
  },
  {
    "count": 951,
    "id": 952,
    "preservationRuleId": 1,
    "word": "word-951",
  },
  {
    "count": 950,
    "id": 951,
    "preservationRuleId": 1,
    "word": "word-950",
  },
  {
    "count": 949,
    "id": 950,
    "preservationRuleId": 1,
    "word": "word-949",
  },
  {
    "count": 948,
    "id": 949,
    "preservationRuleId": 1,
    "word": "word-948",
  },
  {
    "count": 947,
    "id": 948,
    "preservationRuleId": 1,
    "word": "word-947",
  },
  {
    "count": 946,
    "id": 947,
    "preservationRuleId": 1,
    "word": "word-946",
  },
  {
    "count": 945,
    "id": 946,
    "preservationRuleId": 1,
    "word": "word-945",
  },
  {
    "count": 944,
    "id": 945,
    "preservationRuleId": 1,
    "word": "word-944",
  },
  {
    "count": 943,
    "id": 944,
    "preservationRuleId": 1,
    "word": "word-943",
  },
  {
    "count": 942,
    "id": 943,
    "preservationRuleId": 1,
    "word": "word-942",
  },
  {
    "count": 941,
    "id": 942,
    "preservationRuleId": 1,
    "word": "word-941",
  },
  {
    "count": 940,
    "id": 941,
    "preservationRuleId": 1,
    "word": "word-940",
  },
  {
    "count": 939,
    "id": 940,
    "preservationRuleId": 1,
    "word": "word-939",
  },
  {
    "count": 938,
    "id": 939,
    "preservationRuleId": 1,
    "word": "word-938",
  },
  {
    "count": 937,
    "id": 938,
    "preservationRuleId": 1,
    "word": "word-937",
  },
  {
    "count": 936,
    "id": 937,
    "preservationRuleId": 1,
    "word": "word-936",
  },
  {
    "count": 935,
    "id": 936,
    "preservationRuleId": 1,
    "word": "word-935",
  },
  {
    "count": 934,
    "id": 935,
    "preservationRuleId": 1,
    "word": "word-934",
  },
  {
    "count": 933,
    "id": 934,
    "preservationRuleId": 1,
    "word": "word-933",
  },
  {
    "count": 932,
    "id": 933,
    "preservationRuleId": 1,
    "word": "word-932",
  },
  {
    "count": 931,
    "id": 932,
    "preservationRuleId": 1,
    "word": "word-931",
  },
  {
    "count": 930,
    "id": 931,
    "preservationRuleId": 1,
    "word": "word-930",
  },
  {
    "count": 929,
    "id": 930,
    "preservationRuleId": 1,
    "word": "word-929",
  },
  {
    "count": 928,
    "id": 929,
    "preservationRuleId": 1,
    "word": "word-928",
  },
  {
    "count": 927,
    "id": 928,
    "preservationRuleId": 1,
    "word": "word-927",
  },
  {
    "count": 926,
    "id": 927,
    "preservationRuleId": 1,
    "word": "word-926",
  },
  {
    "count": 925,
    "id": 926,
    "preservationRuleId": 1,
    "word": "word-925",
  },
  {
    "count": 924,
    "id": 925,
    "preservationRuleId": 1,
    "word": "word-924",
  },
  {
    "count": 923,
    "id": 924,
    "preservationRuleId": 1,
    "word": "word-923",
  },
  {
    "count": 922,
    "id": 923,
    "preservationRuleId": 1,
    "word": "word-922",
  },
  {
    "count": 921,
    "id": 922,
    "preservationRuleId": 1,
    "word": "word-921",
  },
  {
    "count": 920,
    "id": 921,
    "preservationRuleId": 1,
    "word": "word-920",
  },
  {
    "count": 919,
    "id": 920,
    "preservationRuleId": 1,
    "word": "word-919",
  },
  {
    "count": 918,
    "id": 919,
    "preservationRuleId": 1,
    "word": "word-918",
  },
  {
    "count": 917,
    "id": 918,
    "preservationRuleId": 1,
    "word": "word-917",
  },
  {
    "count": 916,
    "id": 917,
    "preservationRuleId": 1,
    "word": "word-916",
  },
  {
    "count": 915,
    "id": 916,
    "preservationRuleId": 1,
    "word": "word-915",
  },
  {
    "count": 914,
    "id": 915,
    "preservationRuleId": 1,
    "word": "word-914",
  },
  {
    "count": 913,
    "id": 914,
    "preservationRuleId": 1,
    "word": "word-913",
  },
  {
    "count": 912,
    "id": 913,
    "preservationRuleId": 1,
    "word": "word-912",
  },
  {
    "count": 911,
    "id": 912,
    "preservationRuleId": 1,
    "word": "word-911",
  },
  {
    "count": 910,
    "id": 911,
    "preservationRuleId": 1,
    "word": "word-910",
  },
  {
    "count": 909,
    "id": 910,
    "preservationRuleId": 1,
    "word": "word-909",
  },
  {
    "count": 908,
    "id": 909,
    "preservationRuleId": 1,
    "word": "word-908",
  },
  {
    "count": 907,
    "id": 908,
    "preservationRuleId": 1,
    "word": "word-907",
  },
  {
    "count": 906,
    "id": 907,
    "preservationRuleId": 1,
    "word": "word-906",
  },
  {
    "count": 905,
    "id": 906,
    "preservationRuleId": 1,
    "word": "word-905",
  },
  {
    "count": 904,
    "id": 905,
    "preservationRuleId": 1,
    "word": "word-904",
  },
  {
    "count": 903,
    "id": 904,
    "preservationRuleId": 1,
    "word": "word-903",
  },
  {
    "count": 902,
    "id": 903,
    "preservationRuleId": 1,
    "word": "word-902",
  },
  {
    "count": 901,
    "id": 902,
    "preservationRuleId": 1,
    "word": "word-901",
  },
  {
    "count": 900,
    "id": 901,
    "preservationRuleId": 1,
    "word": "word-900",
  },
  {
    "count": 899,
    "id": 900,
    "preservationRuleId": 1,
    "word": "word-899",
  },
  {
    "count": 898,
    "id": 899,
    "preservationRuleId": 1,
    "word": "word-898",
  },
  {
    "count": 897,
    "id": 898,
    "preservationRuleId": 1,
    "word": "word-897",
  },
  {
    "count": 896,
    "id": 897,
    "preservationRuleId": 1,
    "word": "word-896",
  },
  {
    "count": 895,
    "id": 896,
    "preservationRuleId": 1,
    "word": "word-895",
  },
  {
    "count": 894,
    "id": 895,
    "preservationRuleId": 1,
    "word": "word-894",
  },
  {
    "count": 893,
    "id": 894,
    "preservationRuleId": 1,
    "word": "word-893",
  },
  {
    "count": 892,
    "id": 893,
    "preservationRuleId": 1,
    "word": "word-892",
  },
  {
    "count": 891,
    "id": 892,
    "preservationRuleId": 1,
    "word": "word-891",
  },
  {
    "count": 890,
    "id": 891,
    "preservationRuleId": 1,
    "word": "word-890",
  },
  {
    "count": 889,
    "id": 890,
    "preservationRuleId": 1,
    "word": "word-889",
  },
  {
    "count": 888,
    "id": 889,
    "preservationRuleId": 1,
    "word": "word-888",
  },
  {
    "count": 887,
    "id": 888,
    "preservationRuleId": 1,
    "word": "word-887",
  },
  {
    "count": 886,
    "id": 887,
    "preservationRuleId": 1,
    "word": "word-886",
  },
  {
    "count": 885,
    "id": 886,
    "preservationRuleId": 1,
    "word": "word-885",
  },
  {
    "count": 884,
    "id": 885,
    "preservationRuleId": 1,
    "word": "word-884",
  },
  {
    "count": 883,
    "id": 884,
    "preservationRuleId": 1,
    "word": "word-883",
  },
  {
    "count": 882,
    "id": 883,
    "preservationRuleId": 1,
    "word": "word-882",
  },
  {
    "count": 881,
    "id": 882,
    "preservationRuleId": 1,
    "word": "word-881",
  },
  {
    "count": 880,
    "id": 881,
    "preservationRuleId": 1,
    "word": "word-880",
  },
  {
    "count": 879,
    "id": 880,
    "preservationRuleId": 1,
    "word": "word-879",
  },
  {
    "count": 878,
    "id": 879,
    "preservationRuleId": 1,
    "word": "word-878",
  },
  {
    "count": 877,
    "id": 878,
    "preservationRuleId": 1,
    "word": "word-877",
  },
  {
    "count": 876,
    "id": 877,
    "preservationRuleId": 1,
    "word": "word-876",
  },
  {
    "count": 875,
    "id": 876,
    "preservationRuleId": 1,
    "word": "word-875",
  },
  {
    "count": 874,
    "id": 875,
    "preservationRuleId": 1,
    "word": "word-874",
  },
  {
    "count": 873,
    "id": 874,
    "preservationRuleId": 1,
    "word": "word-873",
  },
  {
    "count": 872,
    "id": 873,
    "preservationRuleId": 1,
    "word": "word-872",
  },
  {
    "count": 871,
    "id": 872,
    "preservationRuleId": 1,
    "word": "word-871",
  },
  {
    "count": 870,
    "id": 871,
    "preservationRuleId": 1,
    "word": "word-870",
  },
  {
    "count": 869,
    "id": 870,
    "preservationRuleId": 1,
    "word": "word-869",
  },
  {
    "count": 868,
    "id": 869,
    "preservationRuleId": 1,
    "word": "word-868",
  },
  {
    "count": 867,
    "id": 868,
    "preservationRuleId": 1,
    "word": "word-867",
  },
  {
    "count": 866,
    "id": 867,
    "preservationRuleId": 1,
    "word": "word-866",
  },
  {
    "count": 865,
    "id": 866,
    "preservationRuleId": 1,
    "word": "word-865",
  },
  {
    "count": 864,
    "id": 865,
    "preservationRuleId": 1,
    "word": "word-864",
  },
  {
    "count": 863,
    "id": 864,
    "preservationRuleId": 1,
    "word": "word-863",
  },
  {
    "count": 862,
    "id": 863,
    "preservationRuleId": 1,
    "word": "word-862",
  },
  {
    "count": 861,
    "id": 862,
    "preservationRuleId": 1,
    "word": "word-861",
  },
  {
    "count": 860,
    "id": 861,
    "preservationRuleId": 1,
    "word": "word-860",
  },
  {
    "count": 859,
    "id": 860,
    "preservationRuleId": 1,
    "word": "word-859",
  },
  {
    "count": 858,
    "id": 859,
    "preservationRuleId": 1,
    "word": "word-858",
  },
  {
    "count": 857,
    "id": 858,
    "preservationRuleId": 1,
    "word": "word-857",
  },
  {
    "count": 856,
    "id": 857,
    "preservationRuleId": 1,
    "word": "word-856",
  },
  {
    "count": 855,
    "id": 856,
    "preservationRuleId": 1,
    "word": "word-855",
  },
  {
    "count": 854,
    "id": 855,
    "preservationRuleId": 1,
    "word": "word-854",
  },
  {
    "count": 853,
    "id": 854,
    "preservationRuleId": 1,
    "word": "word-853",
  },
  {
    "count": 852,
    "id": 853,
    "preservationRuleId": 1,
    "word": "word-852",
  },
  {
    "count": 851,
    "id": 852,
    "preservationRuleId": 1,
    "word": "word-851",
  },
  {
    "count": 850,
    "id": 851,
    "preservationRuleId": 1,
    "word": "word-850",
  },
  {
    "count": 849,
    "id": 850,
    "preservationRuleId": 1,
    "word": "word-849",
  },
  {
    "count": 848,
    "id": 849,
    "preservationRuleId": 1,
    "word": "word-848",
  },
  {
    "count": 847,
    "id": 848,
    "preservationRuleId": 1,
    "word": "word-847",
  },
  {
    "count": 846,
    "id": 847,
    "preservationRuleId": 1,
    "word": "word-846",
  },
  {
    "count": 845,
    "id": 846,
    "preservationRuleId": 1,
    "word": "word-845",
  },
  {
    "count": 844,
    "id": 845,
    "preservationRuleId": 1,
    "word": "word-844",
  },
  {
    "count": 843,
    "id": 844,
    "preservationRuleId": 1,
    "word": "word-843",
  },
  {
    "count": 842,
    "id": 843,
    "preservationRuleId": 1,
    "word": "word-842",
  },
  {
    "count": 841,
    "id": 842,
    "preservationRuleId": 1,
    "word": "word-841",
  },
  {
    "count": 840,
    "id": 841,
    "preservationRuleId": 1,
    "word": "word-840",
  },
  {
    "count": 839,
    "id": 840,
    "preservationRuleId": 1,
    "word": "word-839",
  },
  {
    "count": 838,
    "id": 839,
    "preservationRuleId": 1,
    "word": "word-838",
  },
  {
    "count": 837,
    "id": 838,
    "preservationRuleId": 1,
    "word": "word-837",
  },
  {
    "count": 836,
    "id": 837,
    "preservationRuleId": 1,
    "word": "word-836",
  },
  {
    "count": 835,
    "id": 836,
    "preservationRuleId": 1,
    "word": "word-835",
  },
  {
    "count": 834,
    "id": 835,
    "preservationRuleId": 1,
    "word": "word-834",
  },
  {
    "count": 833,
    "id": 834,
    "preservationRuleId": 1,
    "word": "word-833",
  },
  {
    "count": 832,
    "id": 833,
    "preservationRuleId": 1,
    "word": "word-832",
  },
  {
    "count": 831,
    "id": 832,
    "preservationRuleId": 1,
    "word": "word-831",
  },
  {
    "count": 830,
    "id": 831,
    "preservationRuleId": 1,
    "word": "word-830",
  },
  {
    "count": 829,
    "id": 830,
    "preservationRuleId": 1,
    "word": "word-829",
  },
  {
    "count": 828,
    "id": 829,
    "preservationRuleId": 1,
    "word": "word-828",
  },
  {
    "count": 827,
    "id": 828,
    "preservationRuleId": 1,
    "word": "word-827",
  },
  {
    "count": 826,
    "id": 827,
    "preservationRuleId": 1,
    "word": "word-826",
  },
  {
    "count": 825,
    "id": 826,
    "preservationRuleId": 1,
    "word": "word-825",
  },
  {
    "count": 824,
    "id": 825,
    "preservationRuleId": 1,
    "word": "word-824",
  },
  {
    "count": 823,
    "id": 824,
    "preservationRuleId": 1,
    "word": "word-823",
  },
  {
    "count": 822,
    "id": 823,
    "preservationRuleId": 1,
    "word": "word-822",
  },
  {
    "count": 821,
    "id": 822,
    "preservationRuleId": 1,
    "word": "word-821",
  },
  {
    "count": 820,
    "id": 821,
    "preservationRuleId": 1,
    "word": "word-820",
  },
  {
    "count": 819,
    "id": 820,
    "preservationRuleId": 1,
    "word": "word-819",
  },
  {
    "count": 818,
    "id": 819,
    "preservationRuleId": 1,
    "word": "word-818",
  },
  {
    "count": 817,
    "id": 818,
    "preservationRuleId": 1,
    "word": "word-817",
  },
  {
    "count": 816,
    "id": 817,
    "preservationRuleId": 1,
    "word": "word-816",
  },
  {
    "count": 815,
    "id": 816,
    "preservationRuleId": 1,
    "word": "word-815",
  },
  {
    "count": 814,
    "id": 815,
    "preservationRuleId": 1,
    "word": "word-814",
  },
  {
    "count": 813,
    "id": 814,
    "preservationRuleId": 1,
    "word": "word-813",
  },
  {
    "count": 812,
    "id": 813,
    "preservationRuleId": 1,
    "word": "word-812",
  },
  {
    "count": 811,
    "id": 812,
    "preservationRuleId": 1,
    "word": "word-811",
  },
  {
    "count": 810,
    "id": 811,
    "preservationRuleId": 1,
    "word": "word-810",
  },
  {
    "count": 809,
    "id": 810,
    "preservationRuleId": 1,
    "word": "word-809",
  },
  {
    "count": 808,
    "id": 809,
    "preservationRuleId": 1,
    "word": "word-808",
  },
  {
    "count": 807,
    "id": 808,
    "preservationRuleId": 1,
    "word": "word-807",
  },
  {
    "count": 806,
    "id": 807,
    "preservationRuleId": 1,
    "word": "word-806",
  },
  {
    "count": 805,
    "id": 806,
    "preservationRuleId": 1,
    "word": "word-805",
  },
  {
    "count": 804,
    "id": 805,
    "preservationRuleId": 1,
    "word": "word-804",
  },
  {
    "count": 803,
    "id": 804,
    "preservationRuleId": 1,
    "word": "word-803",
  },
  {
    "count": 802,
    "id": 803,
    "preservationRuleId": 1,
    "word": "word-802",
  },
  {
    "count": 801,
    "id": 802,
    "preservationRuleId": 1,
    "word": "word-801",
  },
  {
    "count": 800,
    "id": 801,
    "preservationRuleId": 1,
    "word": "word-800",
  },
]
`;

exports[`preservationRules updateWordCounts successfully updates word counts for a preservation rule 1`] = `
[
  {
    "count": 4,
    "word": "a",
  },
  {
    "count": 1,
    "word": "also",
  },
  {
    "count": 1,
    "word": "definitely",
  },
  {
    "count": 1,
    "word": "hello",
  },
  {
    "count": 1,
    "word": "is",
  },
  {
    "count": 1,
    "word": "maybe",
  },
  {
    "count": 5,
    "word": "message",
  },
  {
    "count": 1,
    "word": "not",
  },
  {
    "count": 6,
    "word": "other",
  },
  {
    "count": 1,
    "word": "test",
  },
  {
    "count": 7,
    "word": "test",
  },
  {
    "count": 3,
    "word": "this",
  },
]
`;
